edx_exams/apps/api/v1/tests/test_views.py

```diff
@@ -513,3 +514,279 @@ def test_proctoring_providers_list_empty(self):
 
         self.assertEqual(response.status_code, 200)
         self.assertEqual(len(response.data), 0)
+
+
+@ddt.ddt
+class ExamAccessTicketsViewsTests(ExamsAPITestCase):
+    """
+    Tests for Exam Access Ticket Views.
+    """
+
+    def setUp(self):
+        super().setUp()
+
+        self.course_id = 'course-v1:edx+test+f19'
+
+        self.due_date = timezone.now() + timedelta(minutes=5)
+        self.exam = Exam.objects.create(
+            resource_id=str(uuid.uuid4()),
+            course_id=self.course_id,
+            provider=self.test_provider,
+            content_id='abcd1234',
+            exam_name='test_exam',
+            exam_type='proctored',
+            time_limit_mins=30,
+            due_date=self.due_date,
+            hide_after_due=False,
+            is_active=True
+        )
+        self.exam_id = self.exam.id
+        self.url = reverse('api:v1:exam-access-tickets', kwargs={'exam_id': self.exam_id})
+
+        self.past_due_date = timezone.now() - timedelta(minutes=5)
+        self.past_due_exam = Exam.objects.create(
+            resource_id=str(uuid.uuid4()),
+            course_id=self.course_id,
+            provider=self.test_provider,
+            content_id='abcd1234',
+            exam_name='test_exam',
+            exam_type='proctored',
+            time_limit_mins=30,
+            due_date=self.past_due_date,
+            hide_after_due=False,
+            is_active=True
+        )
+
+        self.past_due_exam_id = self.past_due_exam.id
+        self.past_due_url = reverse('api:v1:exam-access-tickets', kwargs={'exam_id': self.past_due_exam_id})
+
+    def get_exam_access(self, user, url):
+        """
+        Helper function to make a get request
+        """
+        headers = self.build_jwt_headers(user)
+        return self.client.get(url, **headers)
+
+    def test_auth_failures(self):
+        """
+        Verify the endpoint validates permissions
+        """
+        # Test unauthenticated
+        response = self.client.get(self.url)
+        self.assertEqual(response.status_code, 401)
+
+    def test_exam_not_found(self):
+        """
+        Verify the endpoint returns 404 if exam is not found
+        """
+        url = reverse('api:v1:exam-access-tickets', kwargs={'exam_id': 674})
+
+        headers = self.build_jwt_headers(self.user)
+        response = self.client.get(url, **headers)
+        self.assertEqual(response.status_code, 404)
+
+    def test_access_not_granted(self):
+        """
+        Verify the endpoint doesn't grant access for an exam
+        without an existing exam attempt or past due date.
+        """
+        response = self.get_exam_access(self.user, self.url)
+        self.assertEqual(403, response.status_code)
+
+    @ddt.data(
+        ('started', 200),
+        ('created', 403),
+    )
+    @ddt.unpack
+    def test_access_granted_started_exam_attempt(self, attempt_status, response_status):
```

Comment:
This test and `test_access_granted_verified_attempt_status` appear to be exactly the same other than the exam object. Any reason not to just parameterize exam_past_due as an input and collapse these into a single function?

Reply:
Good point! The reason I didn't is because adding self.exam as a parameter was returning unresolved reference 'self'/NameError: name 'self' is not defined. I did check in the original proctoring example you'd sent, and didn't see any examples of parameterizing objects instantiated in setup. I could create two new exam objects as parameters but I figured this would be redundant which is why I kept as two separate tests. Let me know if there is another way to approach.

Reply:
Yeah, generally you want to keep these parameters to primitives (strings, numbers, booleans) because they actually become part of the test name. (the first test run here is actually `test_access_granted_started_exam_attempt_started_200`) so the way to do this is just pass a boolean for exam_past_due and setup based on that.

eg:
```
def test_access_granted_started_exam_attempt(self, attempt_status, response_status, exam_past_due):
    exam = self.past_due_exam if exam_past_due else self.exam
    etc...
```